swan-lake/development-tutorials/test-document-the-code/ballerina-profiler.md

```diff
@@ -0,0 +1,365 @@
+
+Ballerina Profiler is a tool that monitors Ballerina runtime and its operations such as function calls. It can be used to understand the behavior and troubleshoot the performance issues of a Ballerina program and optimize it.
+
+>**Note:** Ballerina Profiler is an experimental feature, which supports only a limited set of functionality.
+
+## Features
+
+- Profiles a local session of a Ballerina program through offline instrumentation.
+
+- CPU profiling which finds out where the CPU time is going.
+
+- Generating a flame graph which shows the function call stack and the execution times of each function call. It provides an interface for viewing runtime performance and the ability to search, diagnose, and find performance bottlenecks.
+
+>**Note:** Profiling is a high-powered activity with a heavy overhead and can slow down your application.
+
+## Profile a Ballerina program
+
+To profile a Ballerina package, you can use the `bal profile` CLI command inside the root directory of the Ballerina package.
+
+For example, create a Ballerina package named `sort` using the `bal new` CLI command.
+
+```
+$ bal new sort 
+```
+
+Go to the root directory of the Ballerina package.
+
+```
+$ cd sort
+```
+
+Replace the contents of the `main.bal` file with the following Ballerina code, which creates an array of random integers, sorts them, and verifies the output.
+
+```ballerina
+import ballerina/io;
+import ballerina/random;
+
+public function main() returns error? {
+    int[] arr = check createRandomIntArray(check float:pow(10, 2).cloneWithType(int));
+    int[] sortedArr = bubbleSort(arr);
+    boolean isSorted = isSortedArray(sortedArr);
+    io:println("Is the array sorted? " + isSorted.toString());
+}
+
+public isolated function bubbleSort(int[] arr) returns int[] {
+    int n = arr.length();
+    int temp = 0;
+    boolean swapped = false;
+    foreach int i in 0 ... n - 2 {
+        foreach int j in 1 ... n - 1 - i {
+            if (arr[j - 1] > arr[j]) {
+                temp = arr[j - 1];
+                arr[j - 1] = arr[j];
+                arr[j] = temp;
+                swapped = true;
+            }
+        }
+        if (!swapped) {
+            break;
+        }
+    }
+    return arr;
+}
+
+isolated function isSortedArray(int[] sortedArr) returns boolean {
+    foreach int i in 0 ..< sortedArr.length() - 1 {
+        if (sortedArr[i] > sortedArr[i + 1]) {
+            return false;
+        }
+    }
+    return true;
+}
+
+isolated function createRandomIntArray(int size) returns int[]|error {
+    int[] array = [];
+    foreach int i in 0 ..< size {
+        array.push(check random:createIntInRange(0, int:MAX_VALUE));
+    }
+    return array;
+}
+```
+
+Run and profile the Ballerina package using the `bal profile` CLI command.
```

Comment:
```suggestion
Run and profile the Ballerina package using the `bal profile` CLI command.
```
If this is another CLI tool, this should belong to the set of tool pages. Check the format we have followed in them.

Reply:
This is not an integration tool. I discussed with my team to add a separate card saying `Troubleshoot the runtime` under development tutorials, and add the `Ballerina profiler` and `Strand dump tool` there.

Reply:
Edited the below card like this.
<img width="428" alt="image" src="https://github.com/ballerina-platform/ballerina-dev-website/assets/33286289/4915f577-bcfd-43df-8f1d-31cb975ce5af">
Added the below card as follows.
<img width="938" alt="image" src="https://github.com/ballerina-platform/ballerina-dev-website/assets/33286289/5d9d0825-5e68-4dbc-b1c2-0919ea86a5f1">